Fix BufferStore=0 and Ldc != Ldd case

AlexBrownAMD opened the pull request:
When using global store, D offset was being calculated based on C stride. This change adds correct offset calculations for D in BufferStore=0 case.

nakajee commented:
Have you tried LdcEqualsLdd=True?

nakajee commented on Tensile/KernelWriterAssembly.py:
```diff
@@ -12013,7 +12015,9 @@ def globalWriteBatch(self, kernel, ss, batchIdx, applyAlpha, beta, edge, atomic,
           LoadCCodeMod.addCode(LoadCCodeStr)
-
+          
+      if not kernel["BufferStore"] and not kernel["LdcEqualsLdd"]:
```
I think we need to use AssertCEqualsD instead of LdcEqualsLdd.
LdcEqualsLdd seems to be deprecated.

AlexBrownAMD replied:
Ok, we might need to mark as deprecated and reove

AlexBrownAMD replied:
AssertCEqualsD has a different meaning, that's for when C and D are the same memory buffer. LdcEqualsLdd was just for equal stride values, but if it's not working and no longer needed it should probably just be removed.

nakajee replied:
> Ok, we might need to mark as deprecated and reove

It is already written in the definition of LdcEqualsLdd.

nakajee replied:
> AssertCEqualsD has a different meaning, that's for when C and D are the same memory buffer. LdcEqualsLdd was just for equal stride values, but if it's not working and no longer needed it should probably just be removed.

I understand they are different. CEquelsD implies ldc=ldd and we can do same optimizations.
However, that might not be needed.
We can focus on ldc!=ldd only without considering optimization for ldc=ldd at this moment.

nakajee commented on Tensile/KernelWriterAssembly.py:
```diff
@@ -10963,11 +10965,11 @@ def emitAddressSetupCode(self, kernel, ss, tmpVgpr, tmpS01, edge, beta, atomic,
       # calculate global load offset
       # this has to be done after calculating new coord1
-      if not kernel["BufferStore"]:
+      if not kernel["BufferStore"] and (tc == 'D' or (tc == 'C' and beta)):
```
It looks like GLOBAL_OFFSET C or D code is not generated for ldcEqualLdd and not beta case.
We need to change this part like either change this part of condition or call this function with D.

AlexBrownAMD replied:
Updated

nakajee commented on Tensile/KernelWriterAssembly.py:
```diff
@@ -10963,11 +10965,11 @@ def emitAddressSetupCode(self, kernel, ss, tmpVgpr, tmpS01, edge, beta, atomic,
         # global offset macro (requires 3 tmpVgpr)
         # final address = C + index*bytes
-        kStr += "GLOBAL_OFFSET_C %u" % addrVgpr
+        kStr += "GLOBAL_OFFSET_%s %u" % (tc, addrVgpr)
```
It looks like v_mov to tmpVgpr (line 10981,10982 after change) is redundant.
Originally, these v_mov is before setting exec flag and v_add is after exec flag.
With your change, GLOBAL_OFFSET_D is already after setting exec flag.
Then, v_mov to tmpVgpr and v_add (addr += tmpVgpr) are all within exec flag.

I guess the original intention for this part is
   (1) GLOBAL_OFFSET and v_mov to tmpVgpr for all thread
   (2) vadd within exec flag (do not execute vadd for out of range part)

However, it looks like we do not need such consideration.
We can simply add the result of GLOBAL_OFFSET with C/D addr.

nakajee commented:
Some extended tests are failed.
I quickly checked them.
I found 2 issues in your current code.
(1) bpe for GLOBAL_OFFSET_D macro
  we have a condition to use bpe external for C. We need to add tc=="D" case to use proper bpe for D
(2) Shift Pointer for MFMA code
  it looks like we should not generate this code for D. Generating this part of code twice seems not working.

nakajee commented:
Have you tried GSU>1 (+ MultiBuffer(?)) scenario?
It looks like WorkGroup2 calculation is still necessary in non-StridedBatch + GSU case.

nakajee commented on Tensile/KernelWriterAssembly.py:
```diff
@@ -11997,8 +11999,7 @@ def globalWriteBatch(self, kernel, ss, batchIdx, applyAlpha, beta, edge, atomic,
       kStr += addrCalc.emitAddressSetupCode(kernel, ss, 'C', tmpVgpr, tmpS01, edge, beta, atomic, elementIdx, addrDVgpr)
-
-      if edge:
+      if edge and (kernel["BufferStore"] or beta):
```
What about edge + BufferStore + not beta case?
edgeProtectCode will not be called, right?

nakajee commented on Tensile/KernelWriterAssembly.py:
```diff
@@ -12013,6 +12020,11 @@ def globalWriteBatch(self, kernel, ss, batchIdx, applyAlpha, beta, edge, atomic,
+          
+      if not kernel["BufferStore"]:
+        kStr += addrCalc.emitAddressSetupCode(kernel, ss, 'D', tmpVgpr, tmpS01, edge, beta, atomic, elementIdx, addrDVgpr)
+        if edge:
```
We probably do not need to call edgeProtectCode twice for (not BufferStore) + edge + beta case.
It seems not making any difference.

AlexBrownAMD commented:
> Have you tried GSU>1 (+ MultiBuffer(?)) scenario? It looks like WorkGroup2 calculation is still necessary in non-StridedBatch + GSU case.

Fix GSU MB and added new test cases for all combinations including BufferLoad/Store + GSU MB/SB

nakajee commented:
> > Have you tried GSU>1 (+ MultiBuffer(?)) scenario? It looks like WorkGroup2 calculation is still necessary in non-StridedBatch + GSU case.
> 
> Fix GSU MB and added new test cases for all combinations including BufferLoad/Store + GSU MB/SB

Thanks for your update.
I did not clearly say, but according to other location of code, the condition seems to be "not kernel["_GlobalAccumulation"]" rather than kernel["_GlobalAccumulation"] == 'MultipleBuffer'.

We have another option _GlobalAccumulation=SingleBuffer.
This is used in case GSUAlgorithm=SingleBuffer + Dest type is different from Compute type (like HHS).
In either way, as long as extra buffer is allocated for GSU, C,D address seems to be provided as StridedBatch.

nakajee commented:
Please make sure your change works with GSUAlgorithm=SingleBuffer + HHS

AlexBrownAMD commented:
> Thanks for your update. I did not clearly say, but according to other location of code, the condition seems to be "not kernel["_GlobalAccumulation"]" rather than kernel["_GlobalAccumulation"] == 'MultipleBuffer'.

Thanks for the clarification. Updated the checks and added HHS test cases

> Please make sure your change works with GSUAlgorithm=SingleBuffer + HHS + non StridedBatch

Added a test case for this
